### PR TITLE
GH-43964: [Python] Build macOS and manylinux wheels for free-threading

### PR DESCRIPTION
**Pull request by lysnikolaou**
### Rationale for this change

Building free-threaded wheels is necessary to support the free-threaded build. We probably want to upload these wheels as nightlies somewhere as well, so that downstream users can test the free-threading-related changes.

### What changes are included in this PR?

- Add necessary configuration to build 3.13 free-threading wheels on `manylinux`.
- Do necessary changes to build free-threaded wheels on macOS as well.

### Are these changes tested?

I tested the `manylinux` wheel builds. macOS is still untested, since it's not dockerized.

### Are there any user-facing changes?

No.

Related to #43536.
* GitHub Issue: #43964

### CODE REVIEW

**Comment by jorisvandenbossche** [comment]
> We probably want to upload these wheels as nightlies somewhere as well, so that downstream users can test the free-threading-related changes.

We are uploading the nightly wheels the scientific-python-nightly-wheels index like other package do (since yesterday, and we still have our own index on gemfury as well), so this should happen automatically

**Comment by jorisvandenbossche** [comment]
For linux (where it already succeeds to the testing step), we will also need to get a 3.13t build of the test image (`python-wheel-manylinux-test-unittests`)

**Comment by lysnikolaou** [comment]
> For linux (where it already succeeds to the testing step), we will also need to get a 3.13t build of the test image (`python-wheel-manylinux-test-unittests`)

Yeah, I'm working on that as well. My understanding was that both `python-wheel-manylinux-test-unittests` and `python-wheel-manylinux-test-imports` need some changes, but they both depend on the Docker official Python images, which do not have a free-threaded variant. 

It might make sense for me to work on that, too, but for now, I'm trying to see how to best circumvent that. Maybe start with an ubuntu images and install a Python from deadsnakes?

**Comment by jorisvandenbossche** [comment]
> Maybe start with an ubuntu images and install a Python from deadsnakes?

I assume it could be similar as the one you are adding for https://github.com/apache/arrow/pull/43671? (that's also using deadsnakes I think)

**Comment by lysnikolaou** [comment]
That's one option. To basically add a _new_ service that installs `python3.13-nogil` and then uses that to test the wheels, however that would mean that we have to call different test services depending on the python version in `github.linux.yml`. I was trying to avoid that by adjusting the already existing services, but it might be too complicated for no good value.

**Comment by lysnikolaou** [comment]
@jorisvandenbossche Could you rerun the wheel building jobs?

**Comment by jorisvandenbossche** [comment]
@github-actions crossbow submit -g wheel

**Comment by lysnikolaou** [comment]
Not sure why the `install_python` script fails. The failure seems like the `$PYTHON` environment variable is not defined, but I'm pretty sure it is defined in `github.osx.yml`. I've pushed a commit that falls back to explicitly declaring the `$python` bash variable inside the script.

Let's please rerun the jobs. (maybe only the macos ones?)

**Comment by raulcd** [comment]
@github-actions crossbow submit wheel-macos-*

**Comment by lysnikolaou** [comment]
And another `wheel-macos-monterey-cp313-cp313t-*` might be needed. Sorry for the back and forth, but this is difficult to test locally, cause it might mess up my Python installation.

**Comment by raulcd** [comment]
> Sorry for the back and forth, but this is difficult to test locally, cause it might mess up my Python installation.

No worries at all, I understand the need of testing back and forth on CI

**Comment by raulcd** [comment]
@github-actions crossbow submit wheel-macos-monterey-cp313-cp313t-*

**Comment by jorisvandenbossche** [comment]
Based on the wheel file name in the upload step in the arm64 build, it seems it tries to upload a non-free-threaded cp313 wheel (instead of cp313t), but a normal one. Although it certainly did build the cp313t wheel (maybe that is related to why it is failing to upload?)

There is also still a warning about cffi enabling the GIL, so there needs to be a `PYTHON_GIL=0` somewhere?

**Comment by lysnikolaou** [comment]
I tried to look more into this, but couldn't figure out what's going wrong. On my arm64 macOS host, everything is okay.

**Comment by jorisvandenbossche** [comment]
That docker error might have been something temporary? If I look at the latest output above (the github interface is a bit confusing, you have to ensure to select the "latest attempt" because the failing build was restarted), it's actually failing to build pandas later on?

**Comment by jorisvandenbossche** [comment]
It's failing with `No such file or directory: 'meson'`, so somehow the installed meson is not on the path?

File: ci/docker/python-free-threaded-wheel-manylinux-test-unittests.dockerfile
Comment (jorisvandenbossche) [inline ci/docker/python-free-threaded-wheel-manylinux-test-unittests.dockerfile:43]: We might have to add `/arrow-dev/bin` to the path, because I assume that pip will install the `meson` executable there, but it is not automatically found?

Comment (jorisvandenbossche) [reply ci/docker/python-free-threaded-wheel-manylinux-test-unittests.dockerfile:43]: https://mesonbuild.com/Quick-guide.html#troubleshooting

**Comment by lysnikolaou** [comment]
> That docker error might have been something temporary? If I look at the latest output above (the github interface is a bit confusing, you have to ensure to select the "latest attempt" because the failing build was restarted), it's actually failing to build pandas later on?

Oh I actually hadn't seen that. Thanks for the pointer.

I've now changed to dockerfile to add `/arrow-dev/bin` to the `PATH`. Additionally, we still need to figure out what's going wrong with `archery crossbow` trying to upload a non-free-threaded wheel to the nightlies index. I tried to look around but wasn't able to find what's wrong. I can look deeper if you give me a few pointers.

**Comment by lysnikolaou** [comment]
Also can we run something like `wheel-*-cp313t`?

**Comment by jorisvandenbossche** [comment]
> Additionally, we still need to figure out what's going wrong with `archery crossbow` trying to upload a non-free-threaded wheel to the nightlies index. I tried to look around but wasn't able to find what's wrong. I can look deeper if you give me a few pointers.

Looking at the failure above again, I think I was wrong with using "upload", because it was the "Verify uploaded artifacts" step that was failing. But indeed then for arm64 it was for some reason trying to verify the wrong wheel name (without the "t").

If that is still failing now in the latest run (it might also have been some temporary glitch?), you could maybe add some debug prints to `dev/archery/archery/crossbow/reports.py` which is the code that is doing this verification (`ConsoleReport.show()` I think, from the `status` function in `dev/archery/archery/crossbow/cli.py`)

**Comment by lysnikolaou** [comment]
Can we rerun the two macOS wheel jobs? The failures look like temporary glitches because of a timeout trying to download nigthly Cython from `scientific-python-nightly-wheels`.

**Comment by jorisvandenbossche** [comment]
Cython had a bug when building for free-threaded python (https://github.com/cython/cython/issues/6387) for a short time, which we are probably picking up, and hopefully fixed tomorrow when the nightly package is updated.

**Comment by lysnikolaou** [comment]
Cython nightly has been updated. Shall we rerun the wheel jobs?

**Comment by raulcd** [comment]
@github-actions crossbow submit wheel-*-cp313t*

**Comment by lysnikolaou** [comment]
The docker-compose jobs pass locally on my macOS host now. Should be good to try another time.

**Comment by jorisvandenbossche** [comment]
@github-actions crossbow submit wheel-*-cp313t*

File: .env
Comment (jorisvandenbossche) [inline .env:75]:
```suggestion
PYTHON_ABI_TAG=cp39
```

I just merged the PR to drop Python 3.8, so can bump this new variable as well (and there will be some small merge conflicts to solve, sorry!)

File: dev/tasks/python-wheels/github.osx.yml
Comment (jorisvandenbossche) [inline dev/tasks/python-wheels/github.osx.yml:144]:
```suggestion
          arch -{{ arch }} pip install --pre \
```

**Comment by lysnikolaou** [comment]
Fixed the remaining spots as well as some conflicts after your PR was merged. Let's run the jobs again, please.

**Comment by jorisvandenbossche** [comment]
@github-actions crossbow submit wheel-*-cp313t*

**Comment by lysnikolaou** [comment]
Building and testing the free-threaded wheels succeeds now. Only thing remaining is `archery crossbow` and how it tries to validate the wrong wheel on arm64 macOS. I had a look, but I couldn't understand the code around there too much.

**Comment by lysnikolaou** [comment]
Ok, I ran `archery crossbow` through a debugger after following [the instructions](https://arrow.apache.org/docs/dev/developers/crossbow.html) (BTW they're really helpful, but they probably need to be updated for Github Actions as well). Turns out it was just a simple config error. It should be okay now.

**Comment by jorisvandenbossche** [comment]
> following [the instructions](https://arrow.apache.org/docs/dev/developers/crossbow.html) (BTW they're really helpful, but they probably need to be updated for Github Actions as well).

FYI the more recent link is https://arrow.apache.org/docs/dev/developers/continuous_integration/crossbow.html (it's a bit strange that the old page still exists in the dev docs as a leftover from the docs of a few years ago, that is something we should fix), not sure if it is that more up to date though

**Comment by jorisvandenbossche** [comment]
@github-actions crossbow submit wheel-*-cp313t*

**Comment by lysnikolaou** [comment]
The two manylinux arm64 failures look like temporary glitches? Should we rerun them?

**Comment by pitrou** [comment]
I've restarted them.

**Comment by lysnikolaou** [comment]
And once more for `wheel-manylinux-2014-cp313-cp313t-arm64`?

**Comment by lysnikolaou** [comment]
Looks like everything's passing now! So this might be ready to be merged.

File: ci/docker/python-free-threaded-wheel-manylinux-test-unittests.dockerfile
Comment (pitrou) [inline ci/docker/python-free-threaded-wheel-manylinux-test-unittests.dockerfile:43]: I'm curious, Pandas requires meson and ninja nowadays?

File: ci/scripts/install_python.sh
Comment (pitrou) [inline ci/scripts/install_python.sh:66]: Is there a URL reference that you can give in comment for this?

File: dev/tasks/python-wheels/github.linux.yml
Comment (pitrou) [inline dev/tasks/python-wheels/github.linux.yml:62]: Nit
```suggestion
        if: '{{ python_abi_tag }}' != 'cp313t'
```

File: dev/tasks/python-wheels/github.linux.yml
Comment (pitrou) [inline dev/tasks/python-wheels/github.linux.yml:72]: Same here

File: dev/tasks/python-wheels/github.osx.yml
Comment (pitrou) [inline dev/tasks/python-wheels/github.osx.yml:149]: Shouldn't this be only for free-threaded wheels? Regular wheels should work fine with PyPI-originating packages, and we want to ensure that.

File: ci/docker/python-free-threaded-wheel-manylinux-test-unittests.dockerfile
Comment (lysnikolaou) [reply ci/docker/python-free-threaded-wheel-manylinux-test-unittests.dockerfile:43]: Yes, pandas uses meson/ninja to build.

File: dev/tasks/tasks.yml
Comment (pitrou) [inline dev/tasks/tasks.yml:457]: Can you add a reference to https://github.com/numpy/numpy/issues/26157 ?

File: docker-compose.yml
Comment (pitrou) [inline docker-compose.yml:1180]: Is there an upstream ticket we can link to here?

Comment (lysnikolaou) [reply docker-compose.yml]: Added.

File: dev/tasks/tasks.yml
Comment (lysnikolaou) [reply dev/tasks/tasks.yml]: Done.

File: dev/tasks/python-wheels/github.osx.yml
Comment (lysnikolaou) [reply dev/tasks/python-wheels/github.osx.yml]: Right! I changed the step to include a conditional.

**Comment by lysnikolaou** [comment]
Btw, I found out that I can run the jobs myself, using my own repo as a job queue. So here's a couple of jobs after the latest changes to the `github.osx.yml` file (that changed the disabling of build isolation to only happen for free-threaded wheels).

- [github-wheel-macos-monterey-cp313-cp313t-arm64](https://github.com/lysnikolaou/crossbow/actions/runs/10904554381/job/30261266266)
- [github-wheel-macos-monterey-cp313-cp313t-amd64](https://github.com/lysnikolaou/crossbow/actions/runs/10904554383/job/30261266203)

**Comment by pitrou** [comment]
@github-actions crossbow submit wheel*cp312* wheel*cp313*

**Review by pitrou — APPROVED** [review]
LGTM, thanks a lot for this @lysnikolaou

**Comment by pitrou** [comment]
@github-actions crossbow submit -g python

**Comment by pitrou** [comment]
I see this error on a regular Python 3.13 wheel test:
```
Fatal Python error: config_read_gil: Disabling the GIL is not supported by this build
Python runtime state: preinitialized
```
https://github.com/ursacomputing/crossbow/actions/runs/10906089565/job/30266520154#step:17:39

**Comment by lysnikolaou** [comment]
> I see this error on a regular Python 3.13 wheel test:
> 
> ```
> Fatal Python error: config_read_gil: Disabling the GIL is not supported by this build
> Python runtime state: preinitialized
> ```
> 
> https://github.com/ursacomputing/crossbow/actions/runs/10906089565/job/30266520154#step:17:39

Hmm, good catch. Fixed, along with removing `--no-build-isolation` from macOS now that `pandas` wheels for macOS.

**Comment by pitrou** [comment]
@github-actions crossbow submit wheel*cp312* wheel*cp313*

**Comment by lysnikolaou** [comment]
> crossbow submit wheel_cp312_ wheel_cp313_

Do you see the statuses from this somewhere?

**Comment by pitrou** [comment]
> Do you see the statuses from this somewhere?

No, for some reason this specific incantation seems to hang in its GHA job.

**Comment by lysnikolaou** [comment]
Turns out the following construct was indeed needed in `github.linux.yml`:

```
if: |
  '{{ python_abi_tag }}' == 'cp313t'
```

Otherwise a syntax error is reported. Probably something to do with how Github evaluates expressions.

**Comment by pitrou** [comment]
Ahah! Nice find, and sorry for the disruption.

**Comment by pitrou** [comment]
@github-actions crossbow submit wheel*linux*

**Review by pitrou — APPROVED** [review]

**Comment by pitrou** [comment]
@lysnikolaou Do you also plan to submit a PR for Windows wheels?

**Comment by lysnikolaou** [comment]
> @lysnikolaou Do you also plan to submit a PR for Windows wheels?

Yes, I was planning to do so, as soon as NumPy has a Windows wheel for the `cp313t` ABI.

**Comment by pitrou** [comment]
It seems that compiling Pandas' dev version on arm64 can sometimes fail:
https://github.com/ursacomputing/crossbow/actions/runs/10918109629/job/30302833788#step:11:2259

Hopefully they will soon produce wheels? @lysnikolaou

**Comment by lysnikolaou** [comment]
I'm not sure exactly what the plan is about that, but there's definitely challenges to producing `aarch64` wheels. NumPy produces its `aarch64` wheels on CirrusCI, because Github does not provide aarch64 Linux runners to non-entrprise accounts. However, they've indicated that they'll do so by the end of the year, so maybe it makes sense for pandas to wait until then and not introduce another CI service. Either way, I'll open an upstream issue about it.

**Comment by pitrou** [comment]
Ok, hopefully it won't break too often then...

**Comment by pitrou** [comment]
I'm going to merge now.

**Comment by jorisvandenbossche** [comment]
> It seems that compiling Pandas' dev version on arm64 can sometimes fail:

We could potentially also skip installing pandas when testing the arm64 cp313t wheels, if it turns out to be unstable (given it's not a required test dependency)

**Pull request merged**
